packages/core/core-flows/src/customer/steps/validate-customer-account-creation.ts

```diff
@@ -0,0 +1,49 @@
+import { MedusaError, ModuleRegistrationName } from "@medusajs/utils"
+import { createStep, StepResponse } from "@medusajs/workflows-sdk"
+import { CreateCustomerAccountWorkflowInput } from "../workflows"
+
+export const validateCustomerAccountCreationStepId =
+  "validate-customer-account-creation"
+
+export const validateCustomerAccountCreation = createStep(
+  validateCustomerAccountCreationStepId,
+  async (input: CreateCustomerAccountWorkflowInput, { container }) => {
+    const customerService = container.resolve(ModuleRegistrationName.CUSTOMER)
+
+    const { email } = input.customerData
+
+    if (!email) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Email is required to create a customer"
+      )
+    }
+
+    // Check if customer with email already exists
+    const existingCustomers = await customerService.listCustomers({ email })
+
+    if (existingCustomers?.length) {
+      const hasExistingAccount = existingCustomers.some(
+        (customer) => customer.has_account
+      )
+
+      if (hasExistingAccount && input.authIdentityId) {
+        throw new MedusaError(
+          MedusaError.Types.DUPLICATE_ERROR,
+          "Customer with this email already has an account"
+        )
+      }
+
+      if (!hasExistingAccount && !input.authIdentityId) {
+        throw new MedusaError(
+          MedusaError.Types.DUPLICATE_ERROR,
+          "Guest customer with this email already exists"
+        )
+      }
+    }
+
+    input.customerData.has_account = !!input.authIdentityId
```

Comment:
It's a bit misleading that the validation also modifies the input, maybe we do this in a separate step?

Reply:
Yeah, can do!